calyx-backend/src/primitives.rs

```diff
@@ -0,0 +1,115 @@
+//! Primitive instantiations backend for the Calyx compiler.
+//!
+//! Transforms an [`ir::Context`](crate::ir::Context) into a JSON file that
+//! records the unique primitive instantiations in a program.
+//! Usage: -b primitive-inst [-o <OUTPUT_FILE>]
+//! Adapted from resources.rs.
+
+use std::{collections::HashSet, io};
+
+use crate::traits::Backend;
+use calyx_ir as ir;
+use calyx_utils::{CalyxResult, OutputFile};
+use serde::{Deserialize, Serialize};
+
+#[derive(Default)]
+pub struct PrimitiveInstBackend;
+
+impl Backend for PrimitiveInstBackend {
+    fn name(&self) -> &'static str {
+        "resources"
+    }
+
+    /// OK to run this analysis on any Calyx program
+    fn validate(_ctx: &ir::Context) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    /// Don't need to take care of this for this pass
+    fn link_externs(
+        _ctx: &ir::Context,
+        _file: &mut OutputFile,
+    ) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    fn emit(ctx: &ir::Context, file: &mut OutputFile) -> CalyxResult<()> {
+        let main_comp = ctx
+            .components
+            .iter()
+            .find(|comp| comp.name == ctx.entrypoint)
+            .unwrap();
+
+        let primitive_set: &mut HashSet<PrimitiveInst> = &mut HashSet::new();
+
+        gen_primitive_set(ctx, main_comp, primitive_set);
+
+        write_json(primitive_set.clone(), file)?;
+
+        Ok(())
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+struct PrimitiveInst {
+    name: String,
+    params: Vec<PrimitiveParam>,
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+struct PrimitiveParam {
+    param_name: String,
+    param_value: u64,
+}
+
+/// Accumulates a set with each primitive with a given set of parameters
+/// in the program with entrypoint `main_comp`.
+fn gen_primitive_set(
+    ctx: &ir::Context,
+    main_comp: &ir::Component,
+    primitive_set: &mut HashSet<PrimitiveInst>,
+) {
+    for cell in main_comp.cells.iter() {
+        let cell_ref = cell.borrow();
+        match &cell_ref.prototype {
+            ir::CellType::Primitive {
+                name,
+                param_binding,
+                ..
+            } => {
+                let mut curr_params = Vec::new();
+                for (param_name, param_size) in param_binding.iter() {
+                    let param = PrimitiveParam {
+                        param_name: param_name.to_string(),
+                        param_value: *param_size,
+                    };
+                    curr_params.push(param);
+                }
```

Comment:
Not that this matters very much, but this is probably possible to write in a functional way (and eliminate `mut`, in favor of [`collect`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect)). I think it would look something like this (although I haven't tested it):

```rust
let curr_params: Vec<_> = param_binding.iter().map(|param_name, param_size| {
  PrimitiveParam {
    // unchanged
  }
}).collect();
```

Reply:
This is super cool! Thanks for the suggestion :)